src/lib/AccountFinder.ts

```diff
@@ -0,0 +1,58 @@
+import { AccountInfo } from '@/lib/AccountInfo';
+import { NetworkClient } from '@nimiq/network-client';
+import { WalletStore } from '@/lib/WalletStore';
+
+export default class AccountFinder {
+    public static async findAccounts(
+        deriveAccounts: (startIndex: number, count: number) => Promise<Array<{ address: string, keyPath: string }>>,
+        deriveWalletId: (accounts: Array<{ address: string, keyPath: string, balance: number }>) => Promise<string>,
+        defaultLabel: string = 'Standard Account',
+        onIntermediateResult?: (accounts: Array<{ address: string, keyPath: string, balance: number }>) => void,
+    ): Promise<AccountInfo[]> {
+        await NetworkClient.Instance.init(); // initialize NetworkClient if not initialized yet
+        const foundAccounts: Array<{ address: string, keyPath: string, balance: number }> = [];
+        let nextStartIndex = 0;
+        let derivedAccountsPromise = deriveAccounts(nextStartIndex, AccountFinder.MAX_ALLOWED_GAP);
+        nextStartIndex += AccountFinder.MAX_ALLOWED_GAP;
+        while (true) {
+            const derivedAccounts = await derivedAccountsPromise;
+            if (derivedAccounts.length !== AccountFinder.MAX_ALLOWED_GAP) {
+                throw new Error('\'deriveAccounts\' returned wrong number of accounts');
+            }
+
+            // already start deriving next accounts
+            derivedAccountsPromise = deriveAccounts(nextStartIndex, AccountFinder.MAX_ALLOWED_GAP);
+            nextStartIndex += derivedAccounts.length; // by always advancing in blocks of MAX_ALLOWED_GAP it can happen
+            // that we include accounts that are more than MAX_ALLOWED_GAP apart (e.g. one towards the beginning of
+            // the first block, the other one towards the end of the second block). But that's actually not bad.
+
+            // find accounts with a balance > 0
+            // TODO should use transaction receipts combined with balance
+            let didFindAccounts = false;
+            const userFriendlyAddresses = derivedAccounts.map(({ address }) => address);
+            const balances = await NetworkClient.Instance.getBalance(userFriendlyAddresses);
+            for (const account of derivedAccounts) {
+                const balance = balances.get(account.address);
+                if (balance === undefined || balance === 0) continue;
+                foundAccounts.push(Object.assign({}, account, { balance }));
+                didFindAccounts = true;
```

Comment:
The `didFindAccounts` condition actually doesn't honor the `MAX_ALLOWED_GAP` constant, as it makes the lib potentially search up to 39 accounts ahead.

Like so: if in any block of 20 accounts searched, only the very first is detected as active, `didFindAccounts` is true and the loop runs another time, searching 39 accounts further than the last active index.

That's why my previous implementation had both the `nextIndexToDerive` and the `lastActiveIndex`, to enable giving the call to `deriveAccounts` only the relevant number of remaining indices inside the gap.

Reply:
Will probably keep this behavior, as discussed.